### ISSUE TITLE
`GameMaster(blah)` should better be called `create_game`

### ISSUE THREAD

**Issue by Debilski** [ASPP/pelita#433]
create_game(layout=..., teams=..., rounds=...)

might be a clearer API for users than having to create a weird `GameMaster` object. OTOH, since we cannot hide the `GameMaster` this may lead to additional confusion…

**Comment by jni** [comment]
But, mostly yes? I think this would be very valuable. Currently, debugging and profiling in pelita is hard, and these are core skills that we want them to develop during the tournament. Most in our group gave up trying, including the tutors!

**Comment by Debilski** [comment]
Yes. There’ll be an easier way either way. :)

**Comment by otizonaizit** [comment]
Yes, we are going tohave a pelita sprint and we'll address this kind of issues. Just, this specific issue doesn't make sense per se. By the way, since we have introduced a demo of a bot which uses the debugger, I think it is clear now how to use a debugger with pelita and it is not hard at all.
The profiling is a different beast. But I don't see how we can make it better, given the way profilers work in Python. The students of course can keep on use timeit and others manual ways of measuring time. It's not like profiling is impossible.
OTOH, I agree it would be nice to be able to use a proper profiler, but really I don't see how to do it...

**Comment by Debilski** [comment]
> I think it is clear now how to use a debugger with pelita and it is not hard at all.

Depends where you’re coming from. If you’re used to writing code in an IDE (and many people in science around me are) then it is less obvious. (I don’t have a really good solution though atm.)

**Comment by otizonaizit** [comment]
>>If you’re used to writing code in an IDE (and many people in science around me are) then it is less obvious.
>
>That's exactly right. It's also "impossible" to do e.g. `pudb.runcall()` and set a breakpoint from within the debugger.

But we explicitly refrain from teaching IDEs at ASPP! And we want them for once to see how the low-level debugger looks like. It's like teaching git on the command-line. Debugging pelita with the low-level debugger works already good enough. I really think we have more pressant issues to address, instead of serving the hypothetical IDE user. If people are good enough to use the debugger integrated into their IDE, they probably don't need to be our students in the first place.

The workflow we teach with the debugger is to set a breakpoint manually in the code, then enter the debugger and navigate the code with it. This workflow works today with pelita (documented since 2018, so it is already a big improvement). I seriously doubt anything more is needed for our purposes (I am not talking about scientific coding in general, I am just talking about teaching how to use a debugger at ASPP).

**Comment by jni** [comment]
> But we explicitly refrain from teaching IDEs at ASPP!

... Oops. =P

More seriously: in my debugging class, I showed standard pdb, then pudb (have you used it? It's super-amazing), and finally I showed them how IDEs like VSCode let you set breakpoints by clicking next to the code, and step through with rich display of current variables and a console and everything. Honestly I think it's a much better experience than bare pdb, and people should know it exists. pudb is already super awesome and I could just stick with that (`import pudb; pudb.set_trace()`). However, my preference with debugging is *not* to edit the code, but rather run `pudb.runcall(func, arg1, arg2)`, then set the breakpoint from within the debugger. For this, as well as for the IDE, I have no solution with pelita.

Finally, I disagree completely that if they use IDEs they are too advanced. In fact it is quite the opposite: beginners are much more likely to prefer IDEs to vim and the terminal.

Incidentally, I just saw that to set your own debugger with `breakpoint` you need an environment variable. FFS. https://www.andreagrandi.it/2018/10/16/using-ipdb-with-python-37-breakpoint/

Anyway, maybe you are right that this is not top priority. However, there's tags for prioritisation. Priority is not Yes or No. It does not cost anything to have the issue open in case someone who cares deeply about this wants to work on it.

**Comment by Debilski** [comment]
Debugging in Spyder *is* doable with minor (3 lines) changes (runs in the same process now).

<img width="1154" alt="image" src="https://user-images.githubusercontent.com/216179/53211549-df619800-3641-11e9-99d4-94c6bd3ccca8.png">

**Comment by jni** [comment]
👏 👏 👏 

I presume it's the same deal with VSCode, although typically in that case the "run" script (the second part of your file) is a separate file.

Anyway, that's super awesome! Does this mean you'll write a "in_process" flag for `run_game`? or...?

**Comment by Debilski** [comment]
In the demo the code is

    pelita.libpelita.run_game([move, "pelita/player/FoodEatingPlayer.py"], rounds=30, layout=layout)

ie. for the in process players one would just pass the `move` function. External players would still use the same spec strings as one would pass on the command line.

I guess the `run_game` part could also be in a separate file, as long as one can properly import the `move` function it should not be a problem.

A general problem is that there is no good way to make the `in_process` variant work with spec strings (ie. `path/to/group0.py`). Since the Python import logic is so fragile, we can only be sure that the correct thing has been imported when it is not in `sys.modules` *before* we do our import magic (https://github.com/ASPP/pelita/blob/c1fc7058c3fb14c16f2bfafcd81e8d744fe08a3b/pelita/scripts/pelita_player.py#L152) If we didn’t check this then `group0/team/player.py` and `group1/team/player.py` would end up pointing both to `player` in group0 – this is one of the mayor reasons for using a subprocess for each player. This would become even worse in test code.

A pedagogic solution could be to add a wrapper function that explicitly breaks with the symmetry:

    play_my_bot_against(my_move=move, other="blah/player.py")

**Comment by jni** [comment]
@Debilski I'm not looking at the code so I'm not appreciating all the subtleties I'm sure, but isn't `group0` a proper package, and so won't the players necessarily have different module paths? Is there no way to "alias" the imports so that it's clear which is which?

**Comment by Debilski** [comment]
The problem is not aliasing. The problem is that we temporarily change the `sys.path` to include the parent directory of the spec string. Ie. when we have `group0/team/player.py` then `group0/team` gets added to `sys.path` and then we attempt to import `player` (but of course this could be any `player` – that’s why we check that is nonexistent before). It kind of works and most of all it avoids the worst problem of all: The code runs but with a different player (that happens to be in the path somewhere) and you don’t know which.

Of course, we could require that `group0` is a proper module already but then we’d still have to have a way of specifying the root folder to add it to `sys.path`. This would complicate the spec string syntax (`path/to/group0.team.player` ?) and one could still run into a situation where the module has the same name but is simply in a different folder and thus it would be unclear which is run (this is my main problem with installable players in packages).

**Comment by otizonaizit** [comment]
I don't understand. Right now, in pelita_template, we have an __init__.py file in the directory, so the example code we give *already* requires the group directory to be a proper package. I don't see why we need to ask the user to specify the root folder to add it to the path. The root folder can be automatically extracted, *before* importing, by looking at the absolute path of the player.py file, no? And then, you can import and alias the module with from root import player as player_red, for example...

**Comment by Debilski** [comment]
> Am 22.02.2019 um 10:22 schrieb Tiziano Zito <notifications@github.com>:
> 
> I don't understand. Right now, in pelita_template, we have an __init__.py file in the directory, so the example code we give *already* requires the group directory to be a proper package. I don't see why we need to ask the user to specify the root folder to add it to the path. The root folder can be automatically extracted, *before* importing, by looking at the absolute path of the player.py file, no?

That’s what I wrote. That’s how it is done now.

> And then, you can import and alias the module with from root import player as player_red, for example...

I’m not sure if aliasing helps or even makes sense. The `player_red` alias is just a local name and does not affect the import logic.



>

**Comment by otizonaizit** [comment]
>> And then, you can import and alias the module with from root import player as player_red, for example...
>
>I’m not sure if aliasing helps or even makes sense. The `player_red` alias is just a local name and does not affect the import logic.

What problem are we trying to solve then? I don't get it. If we put the root folder in from of sys.path, we are always going to import the right module.

**Comment by Debilski** [comment]
I am not trying to solve anything (apart from the debugging thing). I was just explaining how importing works currently and just mentioned that changing it would introduce more issues than it would solve.
